### ISSUE TITLE
SNOW-1646154: data race in `connection.go`

### ISSUE THREAD

**Issue by lazareviczoran** [snowflakedb/gosnowflake#1203]
Please answer these questions before submitting your issue. 
In order to accurately debug the issue this information is required. Thanks!

1. What version of GO driver are you using?
1.11.0
   
2. What operating system and processor architecture are you using?
Debian 12.5, amd64
   
3. What version of GO are you using?
1.22.5

4.Server version:
8.32.3

5. What did you do?

```go
func list(ctx context.Context, destCh chan Item) {
        db, err := sqlx.ConnectContext(ctx, "snowflake", dsn)
	if err != nil {
		return fmt.Errorf("failed to connect: %w", err)
	}
	defer db.Close()
	
	rows, err := db.QueryxContext(ctx, query)
	if err != nil {
		return fmt.Errorf("failed to run query: %w", err)
	}
	defer rows.Close()
	
	for rows.Next() {
		var item Item
		if err := rows.Scan(&item) {
			return fmt.Errorf("error while fetching items: %w", err)
		}
		// scan and send row to a channel
		destCh <- item
	}
}
```

Above is a simplified function that we use for reading rows from a table, we run a goroutine that connects to snowflake using a context from the main thread and runs the above function. My guess is that when the context gets cancelled and the snowflake connection `Close()`  gets invoked while the `db.QueryxContext` still reads the data and a race has been detected while modifying the `sc.rest` field https://github.com/snowflakedb/gosnowflake/blob/v1.11.0/connection.go#L259.

6. What did you expect to see?

No data race. Potentially shared data should be thread safe.
<details>
  <summary>Race detector output</summary>
  
  ```
  ==================
WARNING: DATA RACE
Write at 0x00c000e85038 by goroutine 194:
  github.com/snowflakedb/gosnowflake.(*snowflakeConn).cleanup()
      /home/zoran/.go/pkg/mod/github.com/snowflakedb/gosnowflake@v1.11.0/connection.go:259 +0xce
  github.com/snowflakedb/gosnowflake.(*snowflakeConn).Close.deferwrap1()
      /home/zoran/.go/pkg/mod/github.com/snowflakedb/gosnowflake@v1.11.0/connection.go:267 +0x1f
  runtime.deferreturn()
      /usr/local/go/src/runtime/panic.go:602 +0x5d
  database/sql.(*driverConn).finalClose.func2()
      /usr/local/go/src/database/sql/sql.go:677 +0x71
  database/sql.withLock()
      /usr/local/go/src/database/sql/sql.go:3530 +0xa6
  database/sql.(*driverConn).finalClose()
      /usr/local/go/src/database/sql/sql.go:675 +0x18d
  database/sql.finalCloser.finalClose-fm()
      <autogenerated>:1 +0x42
  database/sql.(*DB).Close()
      /usr/local/go/src/database/sql/sql.go:930 +0x3fa
  ...

Previous read at 0x00c000e85038 by goroutine 1310:
  github.com/snowflakedb/gosnowflake.downloadChunkHelper()
      /home/zoran/.go/pkg/mod/github.com/snowflakedb/gosnowflake@v1.11.0/chunk_downloader.go:381 +0x38e
  github.com/snowflakedb/gosnowflake.downloadChunk()
      /home/zoran/.go/pkg/mod/github.com/snowflakedb/gosnowflake@v1.11.0/chunk_downloader.go:358 +0x1e2
  github.com/snowflakedb/gosnowflake.(*snowflakeChunkDownloader).schedule.func1()
      /home/zoran/.go/pkg/mod/github.com/snowflakedb/gosnowflake@v1.11.0/chunk_downloader.go:154 +0x90
  github.com/snowflakedb/gosnowflake.init.func1()
      /home/zoran/.go/pkg/mod/github.com/snowflakedb/gosnowflake@v1.11.0/function_wrappers.go:13 +0x26
  github.com/snowflakedb/gosnowflake.(*snowflakeChunkDownloader).schedule.gowrap1()
      /home/zoran/.go/pkg/mod/github.com/snowflakedb/gosnowflake@v1.11.0/chunk_downloader.go:151 +0x61

Goroutine 194 (running) created at:
  golang.org/x/sync/errgroup.(*Group).Go()
      /home/zoran/.go/pkg/mod/golang.org/x/sync@v0.8.0/errgroup/errgroup.go:75 +0x124
  validio.io/ingress/pkg/sources/sourcesnowflake.TestListQueryHistory()
      /home/zoran/services/go/ingress/pkg/sources/sourcesnowflake/query_history_test.go:71 +0x417
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1689 +0x21e
  testing.(*T).Run.gowrap1()
      /usr/local/go/src/testing/testing.go:1742 +0x44

Goroutine 1310 (running) created at:
  github.com/snowflakedb/gosnowflake.(*snowflakeChunkDownloader).schedule()
      /home/zoran/.go/pkg/mod/github.com/snowflakedb/gosnowflake@v1.11.0/chunk_downloader.go:151 +0x31c
  github.com/snowflakedb/gosnowflake.(*snowflakeChunkDownloader).start()
      /home/zoran/.go/pkg/mod/github.com/snowflakedb/gosnowflake@v1.11.0/chunk_downloader.go:141 +0x117b
  github.com/snowflakedb/gosnowflake.(*snowflakeConn).queryContextInternal()
      /home/zoran/.go/pkg/mod/github.com/snowflakedb/gosnowflake@v1.11.0/connection.go:430 +0x74e
  github.com/snowflakedb/gosnowflake.(*snowflakeConn).QueryContext()
      /home/zoran/.go/pkg/mod/github.com/snowflakedb/gosnowflake@v1.11.0/connection.go:366 +0x1e9
  database/sql.ctxDriverQuery()
      /usr/local/go/src/database/sql/ctxutil.go:48 +0xea
  database/sql.(*DB).queryDC.func1()
      /usr/local/go/src/database/sql/sql.go:1776 +0x265
  database/sql.withLock()
      /usr/local/go/src/database/sql/sql.go:3530 +0xa6
  database/sql.(*DB).queryDC()
      /usr/local/go/src/database/sql/sql.go:1771 +0x324
  database/sql.(*DB).query()
      /usr/local/go/src/database/sql/sql.go:1754 +0x168
  database/sql.(*DB).QueryContext.func1()
      /usr/local/go/src/database/sql/sql.go:1732 +0xcf
  database/sql.(*DB).retry()
      /usr/local/go/src/database/sql/sql.go:1566 +0x4a
  database/sql.(*DB).QueryContext()
      /usr/local/go/src/database/sql/sql.go:1731 +0x184
  github.com/jmoiron/sqlx.(*DB).QueryxContext()
      /home/zoran/.go/pkg/mod/github.com/jmoiron/sqlx@v1.4.0/sqlx_context.go:162 +0xbb
  ...
==================
  ```
</details>

7. Can you set logging to DEBUG and collect the logs?

N/A
   
8. What is your Snowflake account identifier, if any? (Optional)
N/A

**Comment by sfc-gh-dszmolka** [comment]
hi - thank you for raising this issue and providing the details ! 
although it is for a somewhat different issue, recently (yesterday :) ) PR https://github.com/snowflakedb/gosnowflake/pull/1196 has been released with gosnowflake 1.11.1 . 
just to rule this one out; do you think it would be possible to upgrade to 1.11.1 and retest ? if it still fails, we'll investigate.

**Comment by lazareviczoran** [comment]
> hi - thank you for raising this issue and providing the details ! although it is for a somewhat different issue, recently (yesterday :) ) PR #1196 has been released with gosnowflake 1.11.1 . just to rule this one out; do you think it would be possible to upgrade to 1.11.1 and retest ? if it still fails, we'll investigate.

Hello, we've noticed the release from yesterday and we've already updated to the latest version. I saw the PR you've referenced and I've looked through the changes and it didn't seem like it would resolve the data race, since the `sc.rest` field is still unguarded, so I decided to open the issue and bring this up. 

I can close the issue for now and and reopen it if we notice this again (or manage to reproduce it).
